File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
public:
    int diameter; // Variable to store the maximum diameter

    int maxDepth(TreeNode* root) {
        if (root == NULL) {
            return 0; // Base case: Null nodes have a depth of 0
        }
        
        // Recursively calculate the left and right subtree depths
        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);
        
        // Update the diameter. The path passing through this node will be the sum of the depths of the left and right subtrees.
        diameter = max(diameter, leftDepth + rightDepth);
        
        // Return the height of the current node
        return 1 + max(leftDepth, rightDepth);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        diameter = 0; // Initialize the diameter
        maxDepth(root); // Calculate the depth while updating the diameter
        return diameter; // Return the maximum diameter
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int dia = 0;
    int height(TreeNode root){
        if(root == null){
            return 0; // Base case: Null nodes have a depth of 0
        }

        // Recursively calculate the left and right subtree depths
        int lh = height(root.left);
        int rh = height(root.right);

        // Update the diameter. The path passing through this node will be the sum of the depths of the left and right subtrees.
        dia = Math.max(dia, lh+rh);

        // Return the height of the current node
        return Math.max(rh,lh)+1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        dia = 0;
        // Calculate the depth while updating the diameter
        height(root);

        return dia;
    }
}